index.bs

```diff
@@ -3631,6 +3635,271 @@ opened</dfn> steps given |window|, |type| and |message|.
+the request, relative to <code>requestTime</code>.
+
+TODO: Add service worker fields
+
+#### The network.Header type #### {#type-network-Header}
+
+[=Remote end definition=] and [=local end definition=]
+
+<pre class="cddl local-cddl">
+NetworkHeader = {
+    name: text,
+    ? value: text,
+    ? binaryValue: [ uint ]
+};
+</pre>
+
+The <code>NetworkHeader</code> type represents a single request header.
+
+If the header value can be represented as a UTF-8 encoded string, the
+<code>value</code> field will be present. Otherwise the <code>binaryValue</code>
+field will be present and consist of an array of integers representing the bytes
+of the header.
+
+#### The network.Initiator type #### {#type-network-Initiator}
+
+[=Remote end definition=] and [=local end definition=]
+
+<pre class="cddl local-cddl">
+NetworkInitiator = {
+    type: "parser" / "script" / "preflight" / "other",
+    ?columnNumber: uint,
+    ?lineNumber: uint,
+    ?stackTrace: StackTrace,
+    ?request: NetworkRequest
+};
+</pre>
+
+The <code>NetworkInitiatior</code> type represents the source of a network request.
+
+TODO: Align more closely with Fetch here?
+
+#### The network.Request type #### {#type-network-RequestId}
+
+[=Remote end definition=] and [=local end definition=]
+
+<pre class="cddl local-cddl remote-cddl">
+NetworkRequest = text;
+</pre>
+
+Each network request has an associated <dfn export>request id</dfn>, which is a string
+uniquely identifying that request.
+
+#### The network.RequestData type #### {#type-network-RequestData}
+
+[=Remote end definition=] and [=local end definition=]
+
+<pre class="cddl local-cddl">
+NetworkRequestData = {
+    url: text,
+    method: text,
+    headers: [ *NetworkHeader ],
+    cookies: [ *NetworkCookie ],
+    ?body: text,
+    bodySize: uint,
+    headersSize: uint
+};
+</pre>
+
+The <code>RequestData</code> type represents an ongoing network request.
+
+TODO: Body is actually bytes, not clear how to handle it as text
+
+#### The network.ResponseData type #### {#type-network-ResponseData}
+
+[=Remote end definition=] and [=local end definition=]
+
+<pre class="cddl local-cddl">
+NetworkResponseData = {
+    url: text,
+    protocol: text,
+    status: unit,
+    statusText: text,
+    headers: [ *NetworkHeader ],
+    cookies: [ *NetworkCookie ],
+    mimeType: text,
+    bytesReceived: uint,
+    timings: NetworkFetchTimingInfo,
+};
+</pre>
+
+### Events ### {#module-network-events}
+
+#### The network.beforeRequestSent Event #### {#event-network-beforeSendRequest}
+<dl>
+   <dt>Event Type</dt>
+   <dd>
+      <pre class="cddl local-cddl">
+        NetworkBeforeRequestSentEvent = {
+         method: "network.beforeRequestSent",
+         params: NetworkBeforeRequestSentParams
+       }
+
+       NetworkBeforeRequestSentParams = {
+        request: NetworkRequest,
+        navigation: Navigation / null,
+        context: BrowsingContext / null,
```

Comment:
For HAR generation, we normally have to map each request to a specific "page" (https://w3c.github.io/web-performance/specs/HAR/Overview.html#pages)

I am wondering if navigation + context will be enough information to match a request with a specific page. Imagine a tab navigates from pageA to pageB. For a late (non-navigation related) request triggered in pageA, we would not get a `navigation` information, only `context`. And since the `context` id remains identical for a top-level browsing context across navigations, we might have to "guess" the page based on when we process this event. This might work but feels racy and I know it's already a complaint users have for HAR creation in Firefox.

Exposing an id unique to the window global could help here, although that's something we would need to also expose in browsing context events?

Reply:
Hmm, yeah, this is a good point. I wonder if we should defer this to after we have support for returning Window objects from scripts; I think that will end up with some kind of per-Window id along with the context id, and maybe we could reuse that here. Or we could use the id of the realm in which the request originated (although arguably that wouldn't make sense for a non-scripted request orginating in a document).

Reply:
We can leave this for later as well. In practice we should be able to infer the page from the time of the request for now (or we can leave the page information out in a first iteration as well)